### PR TITLE
Text changes to all three books for camera updates

### PR DESCRIPTION
**Pull request by hollasch**
An ugly monster of a PR. Probably best to pull this down and review locally.

This should also address #546, #1071, and #464.

### CODE REVIEW

**Comment by trevordblack** [comment]
"and this is what is often called a half-line or ray."
>>
"and this is what is often called a half-line or a ray."

**Comment by trevordblack** [comment]
"that I call ray::at(t):"
>>
"that we'll call ray::at(t):"

File: books/RayTracingInOneWeekend.html [hunk 1]
```diff
@@ -502,54 +502,93 @@
 
 Sending Rays Into the Scene
 ----------------------------
-Now we are ready to turn the corner and make a ray tracer. At its core, a ray tracer sends rays
-through pixels and computes the color seen in the direction of those rays. The involved steps are
+Now we are ready to turn the corner and make a ray tracer.
+At its core, a ray tracer sends rays through pixels and computes the color seen in the direction of
+  those rays.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:507]: ?

File: books/RayTracingInOneWeekend.html [hunk 2]
```diff
@@ -502,54 +502,93 @@
...
 When first developing a ray tracer, I always do a simple camera for getting the code up and running.
 I also make a simple `ray_color(ray)` function that returns the color of the background (a simple
-gradient).
+  gradient).
 
 I’ve often gotten into trouble using square images for debugging because I transpose $x$ and $y$ too
-often, so I’ll use a non-square image.
+  often, so I’ll use a non-square image.
 For now we'll use a 16:9 aspect ratio, since that's so common.
+A 16:9 aspect ratio means that the image width over image height = 16/9.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:521]: "A square image has a 1:1 aspect ratio, because it's width is the same as its height. Since we don't want to use a square image, we'll need to choose a different aspect ratio. We'll choose 16:9 because it's so common. A 16:9 aspect ratio means that the ratio of image width to image height is 16:9. As another example, an image 800 pixels wide by 400 pixels high has a 2:1 aspect ratio"


Consider using latex shenanigans for 16:9 and other aspect ratios

File: books/RayTracingInOneWeekend.html [hunk 3]
```diff
@@ -502,54 +502,93 @@
...
+For example, as an image 800 pixels wide by 400 pixels high has a 2:1 aspect ratio.
+
+We will drive the image (and hence viewport) size from the desired aspect ratio and the image width.
+That way, you can increase or decrease your image size by changing a single value, instead of doing
+  a calculation and changing two values, which you'd have to do if you had to specify both the image
+  width and image height for some target aspect ratio.
+Also, in calculating the image height from the image width and the desired aspect ratio, we'll need
+  to ensure that the resulting height is at least 1.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:529]: "The image's aspect ratio can be determined from the ratio of its height to its width. But since we have a given aspect ratio in mind, it's easier to set the image's width and the aspect ratio, and then using this to calculate for its height. This way, we can scale up or down the image by changing the image width, and it won't throw off our desired aspect ratio. We do have to make sure that when we solve for the image height the resulting height is at least 1."

File: books/RayTracingInOneWeekend.html [hunk 4]
```diff
@@ -502,54 +502,93 @@
...
 
 In addition to setting up the pixel dimensions for the rendered image, we also need to set up a
-virtual viewport through which to pass our scene rays. For the standard pixel spacing (equally
-spaced in both the horizontal and vertical direction), the viewport's aspect ratio should be the
-same as our rendered image. We'll just pick a viewport two units in height. We'll also set the
-distance between the projection plane and the projection point to be one unit. This is referred to
-as the _focal length_, not to be confused with the _focus distance_, which we'll present later.
+  virtual _viewport_ through which to pass our scene rays.
+The viewport is a rectangle in the 3D world that contains the grid of image pixel locations.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:533]: virtual rectangle

File: books/RayTracingInOneWeekend.html [hunk 5]
```diff
@@ -502,54 +502,93 @@
...
+For the standard pixel spacing (equally spaced in both the horizontal and vertical direction), the
+  viewport will have the same aspect ratio as the rendered image.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:535]: If the width is the same as the height of a pixel, the viewport will have the same aspect ratio as the rendered image. The distance between two adjacent pixels is called the pixel spacing, and square pixels is the standard.

**Comment by trevordblack** [comment]
Lots of red ink.

Feel free to liberally interpret my comments

**Comment by trevordblack** [comment]
"To start things off, we'll arbitrarily choose a viewport height of 2.0,"
>>
"To start things off, we'll choose an arbitrary viewport height of 2.0,"

File: books/RayTracingInOneWeekend.html [hunk 6]
```diff
@@ -502,54 +502,94 @@
...
+
+To start things off, we'll arbitrarily choose a viewport height of 2.0, and scale the viewport width
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:537]: we'll choose an arbitrary viewport height of 2.0

File: books/RayTracingInOneWeekend.html [hunk 7]
```diff
@@ -502,54 +502,94 @@
...
+  to give us the desired aspect ratio.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:538]: we'll choose an arbitrary viewport height of 2.0

File: books/RayTracingInOneWeekend.html [hunk 8]
```diff
@@ -502,54 +502,94 @@
...
+Here's a sketch of what this code will look like:
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:539]: Don't use sketch. Conjures image of drawing.

Go with "example" or "snippet" or your choice

File: books/RayTracingInOneWeekend.html [hunk 9]
```diff
@@ -502,54 +502,94 @@
...
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    auto aspect_ratio = 16.0 / 9.0;
+    int image_width = 400;
+
+    // Calculate the image height, and ensure that it's at least 1.
+    int image_height = static_cast<int>(image_width / aspect_ratio);
+    image_height = (image_height < 1) ? 1 : image_height;
+
+    auto viewport_height = 2.0;
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:549]: Consider adding a comment here.:

"// Calculate the viewport height, we don't need to check for 1 because it's a double.

File: books/RayTracingInOneWeekend.html [hunk 10]
```diff
@@ -502,54 +502,94 @@
...
+    auto viewport_width = viewport_height * (static_cast<double>(image_width)/image_height);
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [image-setup]: Rendered image setup]
+
+(If you're wondering why we don't just use `aspect_ratio` when computing `viewport_width`, it's
+because the _actual_ aspect ratio is the integer-based ratio of image width over image height.)
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:555]: ... and the image width was rounded down to the nearest integer, rather than a perfect fraction

File: books/RayTracingInOneWeekend.html (hunk 2 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:521]: ? Do you mean "use the Unicode ratio glyph"? ( `:` vs `∶`)?

File: books/RayTracingInOneWeekend.html [hunk 11]
```diff
@@ -502,54 +502,94 @@
...
+
+Next we will define a _projection point_ in 3D space, from which all scene rays will originate (this
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:557]: In the literature, this is called the "focal point":
https://en.wikipedia.org/wiki/Focal_point
https://en.wikipedia.org/wiki/Focus_(optics)
https://en.wikipedia.org/wiki/Thin_lens

The projection point is the point on a plane nearest to a point.

I'm not sure if projection point is wrong, but it's not what I would go with.

YES, I am aware that focal point is confusing with focal plane, which is a different thing:
https://en.wikipedia.org/wiki/Cardinal_point_(optics)#Focal_planes

Sorry

File: books/RayTracingInOneWeekend.html (hunk 5 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:535]: How about this?

> If pixels are spaced the same distance horizontally as they are vertically, the viewport will have the same aspect ratio as the rendered image.
> The distance between two adjacent pixels is called the pixel spacing, and square pixels is the standard.

File: books/RayTracingInOneWeekend.html (hunk 1 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:507]: Can you elaborate?

File: books/RayTracingInOneWeekend.html (hunk 3 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:529]: I like it.

File: books/RayTracingInOneWeekend.html (hunk 5 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:535]: Better:
> If pixels are spaced the same distance horizontally as they are vertically, the viewport will have the same aspect ratio as the rendered image.
> The distance between two adjacent pixels is called the pixel spacing, and square pixels is the standard.

File: books/RayTracingInOneWeekend.html (hunk 8 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:539]: Done.

File: books/RayTracingInOneWeekend.html (hunk 9 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:549]: Did this:
```
    // Calculate the viewport height. Since it's real valued, we don't check for values less then 1.
```

File: books/RayTracingInOneWeekend.html (hunk 10 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:555]: Updated to
> (If you're wondering why we don't just use `aspect_ratio` when computing `viewport_width`, it's
because the _actual_ aspect ratio is the integer-based ratio of image width over image height, which
will yield zero for values less than one.)

File: books/RayTracingInOneWeekend.html (hunk 1 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:507]: Why is there `  ` sprinkled about at the beginning of lines?

File: books/RayTracingInOneWeekend.html (hunk 9 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:557]: Sorry to nitpick. I prefer
```
// Calculate the viewport height. We don't check for values less than 1 since it's real valued
```

No passive voice in comments

File: books/RayTracingInOneWeekend.html [hunk 12]
```diff
@@ -502,54 +502,103 @@
...
+
+    // Calculate the viewport height. Since it's real valued, we don't check for values less then 1.
+    auto viewport_height = 2.0;
+    auto viewport_width = viewport_height * (static_cast<double>(image_width)/image_height);
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [image-setup]: Rendered image setup]
+
+(If you're wondering why we don't just use `aspect_ratio` when computing `viewport_width`, it's
+because the _actual_ aspect ratio is the integer-based ratio of image width over image height, which
+will yield zero for values less than one.)
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:564]: Are you sure that's correct?

My understanding is that the reason we don't use the `aspect_ratio` is because it's just a user suggestion, and what really matters is the actual ratio of the image dimensions. I don't think that's quite what you're saying here?

File: books/RayTracingInOneWeekend.html [hunk 13]
```diff
@@ -502,54 +502,103 @@
...
+
+I’ll put the camera center at $(0,0,0)$.
 I will have the y-axis go up, and the x-axis to the right.
-In order to respect the convention of a right handed coordinate system, into the screen is the
-negative z-axis.
+Since we use a right-handed coordinate system, the vector into the screen is the negative z-axis.
+Note that I do not make the ray direction a unit length vector because I think not doing that makes
+  for simpler and slightly faster code.
+
+  ![Figure [camera-geom]: Camera geometry](../images/fig-1.03-cam-geom.jpg)
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:578]: Two requests.

1. Can you have a right angle symbol at the intersection of the vector from camera center to  viewport
2. Can you either
  a. Add an intermediate figure between 3 and 4 which is an askew viewport with a grid of pixels on it OR
   b. Add a grid of pixels to the viewport in 3

File: books/RayTracingInOneWeekend.html [hunk 14]
```diff
@@ -502,54 +502,103 @@
...
+    auto viewport_width = viewport_height * (static_cast<double>(image_width)/image_height);
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [image-setup]: Rendered image setup]
+
+(If you're wondering why we don't just use `aspect_ratio` when computing `viewport_width`, it's
+because the _actual_ aspect ratio is the integer-based ratio of image width over image height, which
+will yield zero for values less than one.)
 
-I’ll put the “eye” (or camera center if you think of a camera) at $(0,0,0)$.
+Next we will define the camera center: a point in 3D space from which all scene rays will originate
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:566]: the _camera center_ ?

**Comment by trevordblack** [comment]
Most of the `I`s and `I'll`s should be `We`s and `We'll`s

If it's opinion -> I
If it's something that we're going to do -> We

Should I spin up a separate issue for this?

File: books/RayTracingInOneWeekend.html [hunk 15]
```diff
@@ -502,54 +502,103 @@
...
+  (this is also commonly referred to as the _eye point_).
+The vector from the camera center to the viewport center will be orthogonal to the viewport.
+We'll initially set the distance between the viewport and the camera center point to be one unit.
+This distance is often referred to as the _focal length_.
+
+I’ll put the camera center at $(0,0,0)$.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:572]: "For simplicity we'll start with the camera center at $(0,0,0)$"

File: books/RayTracingInOneWeekend.html (hunk 12 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:564]: Yeah, it hurts my head when I try to parse this. Which doesn't really matter because it's completely bogus. Let me take another swipe.

File: books/RayTracingInOneWeekend.html (hunk 13 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:578]: Ok.

File: books/RayTracingInOneWeekend.html [hunk 16]
```diff
@@ -502,54 +502,103 @@
...
 I will have the y-axis go up, and the x-axis to the right.
-In order to respect the convention of a right handed coordinate system, into the screen is the
-negative z-axis.
+Since we use a right-handed coordinate system, the vector into the screen is the negative z-axis.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:574]: Couple of things

1. I don't think that we need to explain why we're doing y up versus z up, but it's something to consider explaining
2. This is the first time we're mentioned that we're using a right-handed coordinate system. There was no previous mention of what that means, nor was there a previous mention that we're doing this. This reads kind of confusingly. "Since we are doing X"... wait we are? Since when. You need to rephrase this to state that we are in right-handed coordinate systems, and we should also probably say the choice is an arbitrary one

File: books/RayTracingInOneWeekend.html (hunk 14 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:566]: Got it.

File: books/RayTracingInOneWeekend.html (hunk 15 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:572]: I was trying to continue the tone of the original, but it's definitely all over the map throughout the book now.

File: books/RayTracingInOneWeekend.html (hunk 16 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:574]: I would stay far away from the Y-up/Z-up debate. Just tell them what our convention is and move along.

Will reword with respect to RH coordinates.

File: books/RayTracingInOneWeekend.html [hunk 17]
```diff
@@ -502,54 +502,103 @@
...
+Note that I do not make the ray direction a unit length vector because I think not doing that makes
+  for simpler and slightly faster code.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:576]: It's early to state this. We haven't seen any code yet

**Comment by hollasch** [comment]
This is going to be _such_ merge heck.

File: books/RayTracingInOneWeekend.html [hunk 18]
```diff
@@ -502,54 +502,103 @@
...
-$s=0$ is image left, $s=1$ is image right, $t=0$ is image top, and $t=1$ is image bottom.
-That means that we get our vertical axis flip at the right point by computing $t = 1 - y$, where $y$
-is the vertical component of our 3D coordinate system, and $t$ is the flipped vertical component of
-our image coordinate system.
-
-I will traverse the screen from the upper left hand corner, and use two offset vectors along the
-screen sides to move the ray endpoint across the screen.
-Note that I do not make the ray direction a unit length vector because I think not doing that makes
-for simpler and slightly faster code.
+  where the Y axis is inverted (increasing going down).
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:582]: "Now the inevitable tricky part. While our 3D space has the conventions above, this conflicts with our image coordinates, where we want to print out the zeroth pixel in the top-left and work our way down to the last pixel at the bottom-right, meaning that our image coordinates are inverted: An increase in the Y axis goes down. "

File: books/RayTracingInOneWeekend.html [hunk 19]
```diff
@@ -502,54 +502,103 @@
...
+
+As we scan our image, we will start at the upper left pixel (pixel 0,0), scan left-to-right across
+  the row, and then scan the next row, top-to-bottom (reading order for most Latin languages).
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:585]: As we can our image, we will start at the upper-left pixel (pixel \math(0,0)) scan left-to-right across the row, and then scan one row after another from the top to the bottom

I would remove (reading order for most Latin languages)
I'm not sure it's helpful and it's off topic

File: books/RayTracingInOneWeekend.html [hunk 20]
```diff
@@ -502,54 +502,103 @@
...
+To help navigate the pixel grid, we'll use a vector from the left edge to the right edge
+  (`viewport_u`), and a vector from the upper edge to the lower edge (`viewport_v`).
 
-  ![Figure [cam-geom]: Camera geometry](../images/fig-1.03-cam-geom.jpg)
+Our pixel grid will be inset from the viewport edges by half the pixel-to-pixel distance, so each
+  pixel will gather light from the area around it.
+Here's what our viewport and pixel grid look like:
+
+  ![Figure [pixel-grid]: Viewport and pixel grid](../images/fig-1.04-pixel-grid.jpg)
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:593]: The figure should have the pixel annotated P_0,0

File: books/RayTracingInOneWeekend.html [hunk 21]
```diff
@@ -502,54 +502,103 @@
...
-I will traverse the screen from the upper left hand corner, and use two offset vectors along the
-screen sides to move the ray endpoint across the screen.
-Note that I do not make the ray direction a unit length vector because I think not doing that makes
-for simpler and slightly faster code.
+  where the Y axis is inverted (increasing going down).
+
+As we scan our image, we will start at the upper left pixel (pixel 0,0), scan left-to-right across
+  the row, and then scan the next row, top-to-bottom (reading order for most Latin languages).
+To help navigate the pixel grid, we'll use a vector from the left edge to the right edge
+  (`viewport_u`), and a vector from the upper edge to the lower edge (`viewport_v`).
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:587]: have these be math instead of code

File: books/RayTracingInOneWeekend.html [hunk 22]
```diff
@@ -502,54 +502,103 @@
...
 
-  ![Figure [cam-geom]: Camera geometry](../images/fig-1.03-cam-geom.jpg)
+Our pixel grid will be inset from the viewport edges by half the pixel-to-pixel distance, so each
+  pixel will gather light from the area around it.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:590]: "Each Pixel will gather light from the area around it. If pixels were at the grid edge they would gather light from outside of the pixel grid. To avoid this, we will inset the pixels from the viewport edges by half the pixel-to-pixel distance"

Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:590]: Alternively:

""Each Pixel will gather light from the area around it. If pixels were at the grid edge they would gather light from outside of the pixel grid. To avoid this, we will inset the pixels from the viewport edges"

Unless you acknowledge that the sampling area is a pixel in width, saying the inset length may not be needed (or helpful)

File: books/RayTracingInOneWeekend.html [hunk 23]
```diff
@@ -502,54 +502,103 @@
...
+Here's what our viewport and pixel grid look like:
+
+  ![Figure [pixel-grid]: Viewport and pixel grid](../images/fig-1.04-pixel-grid.jpg)
+
+In this figure, we have the viewport, the pixel grid for a 7×5 resolution image, the viewport upper
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:595]: consider \cross instead of ×
if you haven't already (i assume you have, but...)

File: books/RayTracingInOneWeekend.html [hunk 24]
```diff
@@ -502,54 +502,103 @@
...
+  left corner $\mathbf{Q}$, the pixel $\mathbf{P_{0,0}}$ location, the `viewport_u` vector
+  $\mathbf{V_u}$, the `viewport_v` vector $\mathbf{V_v}$, and the pixel delta vectors
+  $\mathbf{\Delta_u}$ and $\mathbf{\Delta_v}$.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:598]: $\mathbf{\Delta u}$ and $\mathbf{\Delta v}$.

File: books/RayTracingInOneWeekend.html [hunk 25]
```diff
@@ -502,54 +502,103 @@
...
-  ![Figure [cam-geom]: Camera geometry](../images/fig-1.03-cam-geom.jpg)
+Our pixel grid will be inset from the viewport edges by half the pixel-to-pixel distance, so each
+  pixel will gather light from the area around it.
+Here's what our viewport and pixel grid look like:
+
+  ![Figure [pixel-grid]: Viewport and pixel grid](../images/fig-1.04-pixel-grid.jpg)
+
+In this figure, we have the viewport, the pixel grid for a 7×5 resolution image, the viewport upper
+  left corner $\mathbf{Q}$, the pixel $\mathbf{P_{0,0}}$ location, the `viewport_u` vector
+  $\mathbf{V_u}$, the `viewport_v` vector $\mathbf{V_v}$, and the pixel delta vectors
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:597]:
```
"  left corner $\mathbf{Q}$, the pixel $\mathbf{P_{0,0}}$ location, the viewport vector
  $\mathbf{V_u}$ (`viewport_v`), the viewport vector $\mathbf{V_v}$ (`viewport_u`), and the pixel delta vectors"
```

File: books/RayTracingInOneWeekend.html [hunk 26]
```diff
@@ -577,18 +626,29 @@
...
+        image_height = (image_height < 1) ? 1 : image_height;
 
         // Camera
 
-        auto viewport_height = 2.0;
-        auto viewport_width = aspect_ratio * viewport_height;
         auto focal_length = 1.0;
+        auto viewport_height = 2.0;
+        auto viewport_width = viewport_height * (static_cast<double>(image_width)/image_height);
+        auto center = point3(0, 0, 0);  // Camera center
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:639]: why not just name it `camera_center`?

File: books/RayTracingInOneWeekend.html [hunk 27]
```diff
@@ -502,54 +502,103 @@
...
+At its core, a ray tracer sends rays through pixels and computes the color seen in the direction of
+  those rays.
+The involved steps are
 
-    1. Calculate the ray from the eye to the pixel,
+    1. Calculate the ray from the “eye” through the pixel,
     2. Determine which objects the ray intersects, and
     3. Compute a color for that intersection point.
 
 When first developing a ray tracer, I always do a simple camera for getting the code up and running.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:514]: Move line 514 down to line 679

File: books/RayTracingInOneWeekend.html [hunk 28]
```diff
@@ -616,9 +677,9 @@
 
 <div class='together'>
 The `ray_color(ray)` function linearly blends white and blue depending on the height of the $y$
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:679]: Don't have the `ray_color` function in the above code block. There's already a lot going on in that.

We also completely dropped the motivation for adding a `ray_color` since we just got inundated with camera stuff. Move line 514-515 to here, and then make a code block for `ray_color`

File: books/RayTracingInOneWeekend.html (hunk 1 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:507]: I'm authoring using semantic linefeeds. Wrapped sentences get indented. This makes it easier to diff text changes  without having to track whole-paragraph reflow. Unfortunately, the reflow issue happens once when going from prose style to semantic linefeeds.

File: books/RayTracingInOneWeekend.html (hunk 27 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:514]: Done.

File: books/RayTracingInOneWeekend.html (hunk 9 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:557]: I don't see any passive voice. It's a combination of command and active. Are you referring to the predicate "since it's real valued"?

Anyway, latest iteration is `// Viewport widths less than one are ok since they are real valued.`

File: books/RayTracingInOneWeekend.html (hunk 11 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html]: Converted to "camera center", per Slack conversation.

File: books/RayTracingInOneWeekend.html (hunk 17 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:576]: Moved to follow the code listing.

File: books/RayTracingInOneWeekend.html (hunk 18 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:582]: Changed to:
> Now the inevitable tricky part.
> While our 3D space has the conventions above, this conflicts with our image coordinates,
>   where we want to have the zeroth pixel in the top-left and work our way down to the last pixel at
>   the bottom right.
> This means that our image coordinate Y-axis is inverted: Y increases going down the image.

File: books/RayTracingInOneWeekend.html (hunk 19 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:585]: Changed to:
> As we scan our image, we will start at the upper left pixel (pixel $0,0$), scan left-to-right across
>   each row, and then scan row-by-row, top-to-bottom.

File: books/RayTracingInOneWeekend.html (hunk 22 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:590]: There's nothing wrong with sampling light outside the pixel grid, or even the viewport. For example, the `sinc` image sampler (one of the best samplers available, though expensive) is infinite in extent. (See https://mathworld.wolfram.com/SincFunction.html.)

But yeah, could use another tweaking. How about this?
> Our pixel grid will be inset from the viewport edges by half the pixel-to-pixel distance.
> This way, our viewport area is evenly divided into width × height identical regions.

File: books/RayTracingInOneWeekend.html (hunk 23 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:595]: This is text instead of math, so I'm using Unicode. Changed to the character entity `&times;`.

File: books/RayTracingInOneWeekend.html (hunk 25 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:597]: Done.

File: books/RayTracingInOneWeekend.html (hunk 12 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:564]: Changed to this:
> Note that `aspect_ratio` is an ideal ratio, which we approximate as best as possible with the
>   integer-based ratio of image width / image height.
> In order for our viewport proportions to exactly match our image proportions, we use the calculated
>   image aspect ratio to determine our final viewport width.

File: books/RayTracingInOneWeekend.html (hunk 16 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:574]: Just reworded to:
> I will have the y-axis go up, the x-axis to the right, and the negative z-axis pointing in the
> viewing direction. (This is commonly referred to as _right-handed coordinates_.)

File: books/RayTracingInOneWeekend.html (hunk 26 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:639]: I liked the succinctness of just `camera` in the code that uses it. And I'd have to rewrap some lines that use it. And the `sphere` class has a member `center`, which is a `<property>` of the `<thing>`.

There — that's all my ammunition.

I'm open to renaming to `camera_center` if you still want; let me know.

**Comment by hollasch** [comment]
Continuing with rebased PR: #1174

File: books/RayTracingInOneWeekend.html (hunk 9 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:555]: I prefer  "We do Y because X" then "since X, we do Y"

File: books/RayTracingInOneWeekend.html (hunk 12 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:562]: Looks good

File: books/RayTracingInOneWeekend.html (hunk 26 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html:645]: While it's in main have it be `camera_center`. Once inside of it's own class rename to `center`?

Comment (hollasch) [reply books/RayTracingInOneWeekend.html:645]: I like that; will do.

File: books/RayTracingInOneWeekend.html (hunk 24 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html]: Done.

File: books/RayTracingInOneWeekend.html [hunk 29]
```diff
@@ -1749,18 +1876,20 @@
...
 
-To handle the multi-sampled color computation, we'll update the `write_color()` function. Rather
-than adding in a fractional contribution each time we accumulate more light to the color, just add
-the full color each iteration, and then perform a single divide at the end (by the number of
-samples) when writing out the color. In addition, we'll add a handy utility function to the
-`interval` class: `clamp(x)`, which clamps the value `x` to the interval:
+First we need to update the `write_color()` function to account for the number of samples we use.
+Rather than adding a fractional contribution each time we accumulate more light to the color, we'll
+  just add the full color from each iteration, and then finish with a single division (by the number
+  of samples) at the end, before writing out the color.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:1889]: "We need to find the average across all of the samples that we take. We'll just add the full color from each iteration, and then finish with a single division (by the number of samples) at the end, before writing out the color."

File: books/RayTracingInOneWeekend.html [hunk 30]
```diff
@@ -1799,15 +1928,48 @@
...
+  overlap each other.
+(See figure 1 of [_Cell Populations of the Retina_](https://iovs.arvojournals.org/article.aspx?articleid=2188127)
+  for an example of the different photoreceptor types in the human retina. There are many more than
+  the single rod + three cone types that most people think of.)
+There's much more to say about this, but if you want to learn more about image sampling and
+  reconstruction, see the excellent technical memo
+  [_A Pixel is Not a Little Square, A Pixel is Not a Little Square, A Pixel is Not a Little Square_](https://www.researchgate.net/publication/244986797),
+  by Alvy Ray Smith, for a fantastic introduction.
+
+In spite of all this, we will proceed with the trivial square pixel model.
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:1952]: From "Which begs the question: _what's the area surrounding the pixel_?" to here.

I'm not sure if the prose makes sense here.
It feels more like Rest of Your Life material than In One Weekend

It's just very deeply technical and is quite a large aside for something that we really don't have the ink to discuss in any amount of detail.

File: books/RayTracingInOneWeekend.html (hunk 9 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:555]: Ok. Well, looks like the above comment satisfies this.

File: books/RayTracingInOneWeekend.html (hunk 15 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:575]: Went with your suggestion.

File: books/RayTracingInOneWeekend.html (hunk 16 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html]: Latest:
> For simplicity we'll start with the camera center at $(0,0,0)$.
> We'll also have the y-axis go up, the x-axis to the right, and the negative z-axis pointing in the
> viewing direction. (This is commonly referred to as _right-handed coordinates_.)

File: books/RayTracingInOneWeekend.html (hunk 29 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:1889]: Done.

File: books/RayTracingInOneWeekend.html (hunk 2 above)
Comment (trevordblack) [reply books/RayTracingInOneWeekend.html]: That or $16:9$

File: books/RayTracingInOneWeekend.html [hunk 31]
```diff
@@ -3085,11 +3267,13 @@
 
 Positionable Camera
 ====================================================================================================
-Cameras, like dielectrics, are a pain to debug. So I always develop mine incrementally. First, let’s
-allow an adjustable field of view (_fov_). This is the angle you see through the portal. Since our
-image is not square, the fov is different horizontally and vertically. I always use vertical fov. I
-also usually specify it in degrees and change to radians inside a constructor -- a matter of
-personal taste.
+Cameras, like dielectrics, are a pain to debug, so I always develop mine incrementally.
+First, let’s allow an adjustable field of view (_fov_).
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html:3271]: allow for an

Comment (hollasch) [reply books/RayTracingInOneWeekend.html:3271]: Done.

**Comment by trevordblack** [comment]
Changing my tact.

I'm going to approve this.
I have always needed to do a walk through and edit for consistency, clarity, simplicity.

I'm going to skim through this for anything I find particularly objectionable and stand by.

Fundamentally the merge could be very painful. It's better to let you do that, and unblock everything else.

File: books/RayTracingInOneWeekend.html (hunk 20 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html:599]: Done.

**Comment by trevordblack** [comment]
A bigger thing is that a substantial portion of the problems I'm noticing are hold-overs from the original text, like 30-40%

**Comment by hollasch** [comment]
Sounds good. I'm going to continue on with a revisit to the image-sampling section to see if I can tighten it up, but as a subsequent PR.

**Comment by trevordblack** [comment]
Looks good for now

**Comment by trevordblack** [comment]
Decide how you want to approach outstanding comments and land

**Review by trevordblack — APPROVED** [review]

**Comment by hollasch** [comment]
*This PR is superseded by PR #1174*
